### ISSUE TITLE
Add GeneralizedLinearRegression with curated family and link support

### ISSUE THREAD

**Issue by drbenvincent** [pymc-labs/CausalPy#1022]
## Problem

CausalPy's native PyMC regression models expose a configurable `y_hat` likelihood through `pymc_extras.prior.Prior`, but the built-in `LinearRegression` graph always constructs `mu = Xβ` with an identity link. Replacing the likelihood alone is therefore insufficient for generalized linear regression: a Poisson `Prior("Poisson")` on `y_hat` still feeds an unconstrained linear predictor into `mu`, which violates both the distribution's support and CausalPy's outcome-scale prediction contract.

The prediction contract (#1016, implemented in #1020) requires that `posterior_predictive["mu"]` be the draw-level conditional expectation `E[Y | parameters, covariates]` on the **response scale**, while `y_hat` remains the noisy likelihood / posterior-predictive variable. Experiments such as `InterruptedTimeSeries` compute causal impact as `y - mu` via `PyMCModel.calculate_impact()`. For count or binary outcomes, `mu` must be the inverse-linked expectation (for example `exp(η)` for Poisson), not the link-scale linear predictor `η`.

Contract tests in `causalpy/tests/test_prediction_contract.py` encode the intended semantics using custom `PyMCModel` subclasses, but there is no public built-in API that wires family, link, and likelihood together. Issue #1017 (customer-support ticket counts ITS tutorial) is blocked on this gap.

## Goal

Add first-class generalized linear regression support for native PyMC regression models through a new `GeneralizedLinearRegression` class, while keeping `LinearRegression` backward compatible as the Gaussian / identity specialization.

Users should be able to fit quasi-experimental models with non-Gaussian outcomes without writing a custom `PyMCModel` subclass, and CausalPy-generated impact, plots, and summaries must remain on the response scale (ticket counts, probabilities, rates).

## Proposed public API

Introduce `causalpy.pymc_models.GeneralizedLinearRegression` with a curated family / link surface:

```python
import causalpy as cp
from causalpy.pymc_models import GeneralizedLinearRegression
from pymc_extras.prior import Prior

model = GeneralizedLinearRegression(
    family="poisson",          # default link: "log"
    # link="log",              # optional override when valid for the family
  sample_kwargs={"random_seed": 42, "progressbar": False},
    priors={
        "beta": Prior("Normal", mu=0, sigma=2.5, dims=["treated_units", "coeffs"]),
        # family-specific auxiliary priors as needed, e.g. "alpha" for NegativeBinomial
    },
)

result = cp.InterruptedTimeSeries(
    data,
    treatment_time,
    formula="tickets ~ 1 + t + C(dow) + active_users",
    model=model,
)
```

### Curated families (initial release)

| `family` | Default `link` | Response parameter passed to likelihood | Notes |
|----------|----------------|----------------------------------------|-------|
| `"gaussian"` | `"identity"` | `mu` | Equivalent to current `LinearRegression` behavior; supports existing `y_hat` `Prior("Normal", sigma=...)` customization |
| `"poisson"` | `"log"` | `mu` | Count outcomes; `mu = exp(eta)` must be strictly positive |
| `"negative_binomial"` | `"log"` | `mu` | Overdispersed counts; expose dispersion via `priors["alpha"]` or equivalent documented default |
| `"bernoulli"` | `"logit"` | `p` | Binary outcomes; graph still exposes outcome-scale `mu` as the probability for impact / contract compliance |

Supported `link` literals for v1: `"identity"`, `"log"`, `"logit"`. Validate family / link pairings at construction time (reject invalid combinations rather than failing silently at sampling). Default `link=None` means use the family canonical link.

`LinearRegression` remains the existing public class and default backend for experiments that do not pass an explicit model. Internally it should delegate to the shared generalized regression implementation (or subclass it) so behavior does not drift.

### Required model graph

For all supported families, `build_model()` must follow the prediction contract:

1. Compute the linear predictor `eta = X @ beta` (optionally exposed as a `Deterministic` named `eta`).
2. Apply the configured inverse link draw by draw to obtain the response-scale expectation.
3. Register that quantity as `pm.Deterministic("mu", ..., dims=["obs_ind", "treated_units"])`.
4. Construct the family likelihood as `y_hat`, using `Prior.create_likelihood_variable` when the distribution accepts a `mu` parameter and a documented branch when it requires `p` (Bernoulli).

`PyMCModel.predict()` must continue to return both `posterior_predictive["mu"]` and `posterior_predictive["y_hat"]` with dims `(chain, draw, obs_ind, treated_units)`.

### Shared regression head

Extract a small internal helper used by `GeneralizedLinearRegression` (and reused by `LinearRegression`) rather than duplicating graph construction. Suggested responsibilities:

- `inverse_link(eta, link)` for `"identity"`, `"log"`, and `"logit"` using PyMC / PyTensor ops (`pt.exp`, `pm.math.sigmoid` / `invlogit`).
- `build_regression_head(X, beta, link) -> (eta, mu)` returning link-scale and response-scale tensors with correct dims.
- `create_family_likelihood(family, mu_or_p, y_obs, priors)` handling `mu`-parameterized and `p`-parameterized distributions.

Keep the surface minimal: no Bambi dependency, no public custom `Family` objects in v1, no statsmodels GLM wrapper.

### Prior customization

Preserve the existing `priors=` dict pattern:

- `"beta"` (and any other coefficient priors) as today.
- `"y_hat"` overrides for Gaussian ancillary parameters (`sigma`) where applicable.
- Documented auxiliary keys per family (for example `"alpha"` for NegativeBinomial).

### Clone / refit

Sensitivity checks (`PlaceboInTime`, `PriorSensitivity`, etc.) clone models via `_clone()`. `GeneralizedLinearRegression._clone()` must copy `family`, `link`, `sample_kwargs`, and `_user_priors` so refits preserve the configured GLM.

### `score()` behavior

`PyMCModel.score()` currently computes Bayesian R² from `mu` and is meaningful for continuous Gaussian outcomes. For non-Gaussian families, either document the limitation, emit a targeted warning, or gate / skip scoring. Do not return silently misleading R² values for Bernoulli or count models.

### In scope

- `GeneralizedLinearRegression` with curated families and links listed above.
- Refactor `LinearRegression` to share the generalized implementation without changing default behavior.
- Public export from `causalpy.pymc_models` and `causalpy` (subject to existing export checks).
- Knowledgebase / API documentation cross-linked from `prediction-contract.md`.
- Fast pytest coverage:
  - Unit tests for link helper and family / link validation.
  - Contract tests mirroring `test_prediction_contract.py` for each curated family on the **built-in** class (not only test subclasses).
  - At least one `InterruptedTimeSeries` integration test with a count outcome showing outcome-scale impact recovery within posterior uncertainty.
- Unblocks #1017 once merged.

### Out of scope (explicit exclusions)

- Adding `family=` / `link=` kwargs directly on experiment classes (`InterruptedTimeSeries`, `DifferenceInDifferences`, etc.). Users configure the model object; experiments stay agnostic.
- Synthetic-control weight models (`WeightedSumFitter`, `SoftmaxWeightedSumFitter`): donor-weighted sums are already outcome-scale counterfactuals; a global outcome link would change the estimand.
- `SyntheticDifferenceInDifferencesWeightFitter`, `PropensityScore`, `InstrumentalVariableRegression`: different graph structures and estimands.
- `BayesianBasisExpansionTimeSeries`, `StateSpaceTimeSeries`: Gaussian observation models with component-specific structure; separate follow-up if non-Gaussian observation models are needed.
- `PyMCForecastModel` / `pymc-forecast` backend: link must be applied inside the user `model_fn`; coordinate via docs and upstream issues (`pymc-forecast#52`, `#50`).
- scikit-learn backends: no `eta` / `mu` / `y_hat` contract; separate design if sklearn GLM support is ever desired.
- Public custom `Family` / `Link` objects, probit / cloglog / softplus links, and zero-inflated families.
- Bambi or statsmodels as dependencies.

## Acceptance criteria

- [ ] `GeneralizedLinearRegression(family="poisson")` builds a graph with `eta`, inverse-linked outcome-scale `mu`, and Poisson `y_hat`.
- [ ] `GeneralizedLinearRegression(family="bernoulli")` exposes probability-scale `mu` and Bernoulli `y_hat` via `p`.
- [ ] `LinearRegression` behavior is unchanged for existing callers (identity Gaussian default).
- [ ] `predict()` returns both `mu` and `y_hat` with dims `(chain, draw, obs_ind, treated_units)`.
- [ ] `calculate_impact()` produces response-scale contrasts (`y - mu`), verified by tests for each curated family.
- [ ] Invalid family / link combinations raise clear errors at model construction.
- [ ] `_clone()` preserves `family` and `link` for sensitivity-check refits.
- [ ] `InterruptedTimeSeries` integration test with a simulated count outcome recovers known effects on the count scale.
- [ ] Documentation updated; `prek run --all-files` passes.

## Related

- #1016 — prediction contract documentation
- #1020 — prediction contract implementation (merged)
- #1017 — customer-support ticket counts ITS tutorial (blocked on this issue)
- `causalpy/tests/test_prediction_contract.py` — semantic ground truth for `eta`, `mu`, and `y_hat`
- `docs/source/knowledgebase/prediction-contract.md` — author-facing contract

## Suggested implementation order

1. Internal `inverse_link` helper + family / link validation module.
2. `GeneralizedLinearRegression.build_model()` for Poisson (proves the full contract path).
3. Refactor `LinearRegression` to share the regression head (identity Gaussian).
4. Add NegativeBinomial and Bernoulli branches.
5. Integration test via `InterruptedTimeSeries`.
6. Documentation and export wiring.
7. Resume #1017 notebook work as the first end-to-end consumer.

**Comment by NathanielF** [comment]
This looks like a good move i think. Will have a proper review this coming week. First impression though, I wonder to what degree it rebuilds some of Bambi's functionality?

**Comment by drbenvincent** [comment]
@NathanielF I might try to move it along a bit, but would be grateful for a review. One thing we've got is a _partial_ switch to estimates through g-computation vs coefficients. I'll try to make that boundary more principled and clear
